seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/source/split/DorisSourceSplitEnumerator.java

```diff
@@ -0,0 +1,191 @@
+        return this.pendingSplit.size();
+    }
+
+    @Override
+    public void handleSplitRequest(int subtaskId) {
+        throw new DorisConnectorException(
+                CommonErrorCodeDeprecated.UNSUPPORTED_OPERATION,
+                String.format("Unsupported handleSplitRequest: %d", subtaskId));
+    }
+
+    @Override
+    public void registerReader(int subtaskId) {
+        log.debug("Register reader {} to DorisSourceSplitEnumerator.", subtaskId);
+        if (!pendingSplit.isEmpty()) {
+            synchronized (stateLock) {
+                assignSplit(Collections.singletonList(subtaskId));
+            }
+        }
+    }
+
+    @Override
+    public DorisSourceState snapshotState(long checkpointId) {
+        synchronized (stateLock) {
+            return new DorisSourceState(shouldEnumerate, pendingSplit);
+        }
+    }
+
+    @Override
+    public void notifyCheckpointComplete(long checkpointId) {}
+
+    private List<DorisSourceSplit> getDorisSourceSplit() {
+        List<DorisSourceSplit> splits = new ArrayList<>();
+        List<PartitionDefinition> partitions =
+                RestService.findPartitions(seaTunnelRowType, dorisConfig, log);
+        for (PartitionDefinition partition : partitions) {
+            splits.add(new DorisSourceSplit(partition, String.valueOf(partition.hashCode())));
+        }
+        return splits;
+    }
+
+    private void addPendingSplit(Collection<DorisSourceSplit> splits) {
+        int readerCount = context.currentParallelism();
+        for (DorisSourceSplit split : splits) {
+            int ownerReader = getSplitOwner(split.splitId(), readerCount);
+            log.info("Assigning {} to {} reader.", split.splitId(), ownerReader);
+            pendingSplit.computeIfAbsent(ownerReader, f -> new ArrayList<>()).add(split);
+        }
+    }
+
+    private static int getSplitOwner(String tp, int numReaders) {
+        return (tp.hashCode() & Integer.MAX_VALUE) % numReaders;
+    }
+
+    private void assignSplit(Collection<Integer> readers) {
+        for (Integer reader : readers) {
+            final List<DorisSourceSplit> assignmentForReader = pendingSplit.remove(reader);
+
+            if (assignmentForReader != null && !assignmentForReader.isEmpty()) {
+
+                String splitsInfo =
+                        assignmentForReader.stream()
+                                .map(DorisSourceSplit::getSplitId)
+                                .collect(Collectors.joining(","));
+                log.info("Assign splits {} to reader {}", splitsInfo, reader);
+                try {
+                    context.assignSplit(reader, assignmentForReader);
+                } catch (Exception e) {
+                    log.error(
+                            "Failed to assign splits {} to reader {}",
+                            assignmentForReader,
+                            reader,
+                            e);
+                    pendingSplit.put(reader, assignmentForReader);
```

Comment:
This will cause split lost.
`readers.forEach(context::signalNoMoreSplits);` will be call after method `assignSplit` return. But may be some split added to `pendingSplit ` because `context.assignSplit(reader, assignmentForReader);` throw Exception.
```
public void run() {
        Set<Integer> readers = context.registeredReaders();
        if (shouldEnumerate) {
            List<DorisSourceSplit> dorisSourceSplits = getDorisSourceSplit();
            synchronized (stateLock) {
                addPendingSplit(dorisSourceSplits);
                shouldEnumerate = false;
                assignSplit(readers);
            }
        }

        log.debug(
                "No more splits to assign." + " Sending NoMoreSplitsEvent to reader {}.", readers);
        readers.forEach(context::signalNoMoreSplits);
    }
```